PyInstaller/building/imphook.py

```diff
@@ -12,18 +12,453 @@
+            self._sanitize_hook_attrs()
+
+        # Return the previously loaded in-memory module.
+        return self._scope
+
+
+    def _sanitize_hook_attrs(self):
+        """
+        Sanitize all top-level attributes defined by this hook script.
+
+        This method converts both the `binaries` and `datas` hook attributes
+        from lists of hook-style 2-tuples into sets of `TOC`-style 2-tuples, as
+        well as defaulting undefined hook attributes to empty lists or sets.
+
+        See Also
+        ----------
+        `format_binaries_and_datas()`
+            Further details on hook- to `TOC`-style conversions.
+        """
+
+        # Convert lists of hook-style 2-tuples into sets of TOC-style 2-tuples,
+        # defaulting undefined lists to empty sets. This must be done before the
+        # subsequently called _process_hook_func() method appends already
+        # converted TOC-style 2-tuples to these sets.
+        for attr_name in ('binaries', 'datas'):
+            setattr(    self.scope, attr_name, set(format_binaries_and_datas(
+                getattr(self.scope, attr_name, list()))))
+
+        # Convert list attributes in which order is insignificant to sets,
+        # defaulting undefined lists to empty sets.
+        for attr_name in ('excludedimports',):
+            setattr(    self.scope, attr_name, set(
+                getattr(self.scope, attr_name, list())))
+
+        # Default all remaining undefined lists to empty. Since order is
+        # significant in these lists, these lists must not be converted to sets.
+        for attr_name in ('hiddenimports',):
+            if  getattr(self.scope, attr_name, None) is None:
+                setattr(self.scope, attr_name, list())
+
+    ## Hooks
+
+    def post_graph(self):
+        """
+        Call the **post-graph hook** (i.e., `hook()` function) defined by this
+        hook script if any _or_ noop otherwise.
+
+        This method is intended to be called _after_ constructing the module
+        graph for this application.
+        """
+
+        # This call modifies hook attributes (e.g., `hiddenimports`) accessed by
+        # other processing methods and hence is called first.
+        self._process_hook_func()
+
+        # Order is insignificant here.
+        self._process_hidden_imports()
+        self._process_excluded_imports()
+
+
+    def _process_hook_func(self):
+        """
+        Call this hook's `hook()` function if defined _or_ noop otherwise.
+        """
+
+        # If this hook script defines no hook() function, noop.
+        if not hasattr(self.scope, 'hook'):
+            return
+
+        # Call this hook() function with a communications API.
+        hook_api = PostGraphAPI(
+            module_name=self.module_name, module_graph=self.module_graph)
+        self.scope.hook(hook_api)
+
+        # Process all attributes of this API modified by the prior call.
+        self.scope.datas.update(set(hook_api._added_datas))
```

Comment:
Yup! Isn't life grand?

The `_sanitize_hook_attrs()` method has already converted the `datas` attribute for this hook from a list of hook-style 2-tuples into a set of `TOC`-style 2-tuples. The above statement then updates that set with the additional `datas` added by this hook's `hook()` function (if any).